Bluetooth: Exclude from build code non related to chosen Bluetooth role

MariuszSkamra opened the pull request:
I made some static analysis of Bluetooth host code and I found few places in which the code can be reduced. The aim of this PR is to exclude from build code non related to chosen Bluetooth role.
This results in image size reduction.
eg. smp_svr on nRF52840; prj_tiny.conf
Before:
Memory region         Used Size  Region Size  %age Used
           FLASH:       91592 B       420 KB     21.30%
            SRAM:       21172 B       256 KB      8.08%
        IDT_LIST:         140 B         2 KB      6.84%

After:
Memory region         Used Size  Region Size  %age Used
           FLASH:       90672 B       420 KB     21.08%
            SRAM:       21172 B       256 KB      8.08%
        IDT_LIST:         140 B         2 KB      6.84%

Almost 1K flash used less

MariuszSkamra committed:
Bluetooth: hci_core: Exclude conn creation related code for non-central

Don't build the code that is related to central role if the target
does not support connection creation.
This helps to reduce the application image size that implement
either Observer, Peripheral or Broadcaster role.

Note:
BT_CENTRAL implies BT_CONN and BT_OBSERVER

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: hci_core: Exclude scan related code if non-observer

This chunks should be excluded from build if the target application
does not support neither observer not central role.
This helps to reduce the application image size that implement
Peripheral or Broadcaster role.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: att: Do not build Signed Write cmd handler if SMP disabled

This function does nothing actually if SMP is disabled.
This excludes att_signed_write_cmd handler from build if BT_SMP
is disabled or BT_SIGNING is disabled.

Note:
BT_SIGNING depends on BT_SMP

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: att: Add Kconfig option to disable Multiple Read operation

This adds an Kconfig option to disable GATT Multiple Read support.
This GATT sub-procedure is optional and does not have to be supported.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
samples: smp_svr: Disable GATT Multiple Read

This feature is not used by mcu_mgr.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented:
Actually, there's more we can disable. If SMP is disabled, Address Resolution, resolving list related stuff can be excluded from build as well. I'll push another version

MariuszSkamra commented:
@cvinayak The same applies to controller, right? If SMP is disabled there is no need to build rl stuff.
Thanks @sjanc for this clue!

carlescufi commented:
@MariuszSkamra brilliant, thanks for this. Could you also do the same for the controller?

jhedberg commented on subsys/bluetooth/host/mesh/Kconfig:
```diff
@@ -11,6 +11,7 @@ menuconfig BT_MESH
 	select TINYCRYPT
 	select TINYCRYPT_AES
 	select TINYCRYPT_AES_CMAC
+	select BT_OBSERVER
```
Most mesh app currently have this in their `prj.conf`:

```
CONFIG_BT_OBSERVER=y
CONFIG_BT_BROADCASTER=y
```

If you're going to add this `select` statement, shouldn't you also add `BROADCASTER`? And if you add these, shouldn't you remove them from the respective `prj.conf` files?

MariuszSkamra replied:
OK, I'll add another patch for this

jhedberg replied:
That said, is it the right thing to add these? There's a general aversion to using select, and it should be limited to very specific cases. @carlescufi do you think this qualifies?

jhedberg replied:
@MariuszSkamra what exactly was it that prompted you to add the select statement? It seems there's a bug in the Kconfig either way: if we don't have select there it's then at least missing a depends on for both observer and broadcaster.

MariuszSkamra replied:
@jhedberg I believe, one of the samples failed when I ran sanitycheck.

MariuszSkamra replied:
I'm running to see which one.
Nevertheless, I think that having those selected in Kconfig is logical. Mesh device is broadcaster and observer imo. Having GATT bearer enforces peripheral role, provisioner over GATT bearer enforces central role etc.

MariuszSkamra replied:
@jhedberg 
`
CONFIG_BT_OBSERVER=y
CONFIG_BT_BROADCASTER=y
`
is missing in samples/bluetooth/mesh_demo/prj.conf

jhedberg replied:
@MariuszSkamra for the record, the rationale for preferring `depends on` and avoiding `select` (and when exceptions can be made) is outlined here:

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/kbuild/kconfig-language.txt#n128

carlescufi replied:
Yep, I completely agree with @jhedberg here. We should add a `depends on` for both instead of the current `select` statements. I see no justification in this case to use `select` other than convenience for the user, but this might backfire later as `select` is dangerous long-term.

MariuszSkamra replied:
@jhedberg @carlescufi Thanks, I see now. You have right, the dependency is the right way to do it. I was not aware that 'select' don't resolve the dependencies.

jhedberg commented:
Btw, this will likely conflict with #8560 so either that or this one will eventually need to be rebased and get the conflicts sorted out

jhedberg commented:
@MariuszSkamra #8560 has now been merged, so you'll want to check that the new API falls behind the right ifdefs that you're adding.

MariuszSkamra committed:
Bluetooth: Mesh: Depend Mesh upon Observer and Broadcaster roles

This adds dependency upon BT_BROADCASTER and BT_OBSERVER roles.
Those are necessary to implement Mesh device.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: mesh_demo: Enable missing options in configuration file

This enables missing Observer and Brodcaster roles in configuration
file.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented:
@jhedberg I didn't have to resolve any conflict ^^ I pushed rebased version with the change in Kconfig.
